ethcore/src/spec/spec.rs

```diff
@@ -21,55 +21,51 @@ use engine::*;
 use pod_state::*;
 use null_engine::*;
 use account_db::*;
+use super::genesis::Genesis;
+use super::seal::Generic as GenericSeal;
 use ethereum;
-use super::genesis::{Seal as GenesisSeal, Genesis};
-
-/// Convert JSON value to equivalent RLP representation.
-// TODO: handle container types.
-fn json_to_rlp(json: &Json) -> Bytes {
-	match *json {
-		Json::Boolean(o) => encode(&(if o {1u64} else {0})).to_vec(),
-		Json::I64(o) => encode(&(o as u64)).to_vec(),
-		Json::U64(o) => encode(&o).to_vec(),
-		Json::String(ref s) if s.len() >= 2 && &s[0..2] == "0x" && U256::from_str(&s[2..]).is_ok() => {
-			encode(&U256::from_str(&s[2..]).unwrap()).to_vec()
-		},
-		Json::String(ref s) => {
-			encode(s).to_vec()
-		},
-		_ => panic!()
-	}
+use ethjson;
+
+/// Parameters common to all engines.
+#[derive(Debug, PartialEq, Clone)]
+pub struct CommonParams {
+	/// Account start nonce.
+	pub account_start_nonce: U256,
+	/// Frontier compatibility mode limit.
+	pub frontier_compatibility_mode_limit: u64,
+	/// Maximum size of extra data.
+	pub maximum_extra_data_size: usize,
+	/// Network id.
+	pub network_id: U256,
+	/// Minimum gas limit.
+	pub min_gas_limit: U256,
 }
 
-/// Convert JSON to a string->RLP map.
-fn json_to_rlp_map(json: &Json) -> HashMap<String, Bytes> {
-	json.as_object().unwrap().iter().map(|(k, v)| (k, json_to_rlp(v))).fold(HashMap::new(), |mut acc, kv| {
-		acc.insert(kv.0.clone(), kv.1);
-		acc
-	})
+impl From<ethjson::spec::Params> for CommonParams {
+	fn from(p: ethjson::spec::Params) -> Self {
+		CommonParams {
+			account_start_nonce: p.account_start_nonce.into(),
+			frontier_compatibility_mode_limit: p.frontier_compatibility_mode_limit.into(),
+			maximum_extra_data_size: p.maximum_extra_data_size.into(),
+			network_id: p.network_id.into(),
+			min_gas_limit: p.min_gas_limit.into(),
+		}
+	}
 }
 
 /// Parameters for a block chain; includes both those intrinsic to the design of the
 /// chain and those to be interpreted by the active chain engine.
-#[derive(Debug)]
 pub struct Spec {
 	/// User friendly spec name
 	pub name: String,
 	/// What engine are we using for this?
-	pub engine_name: String,
+	pub engine: Box<Engine>,
```

Comment:
`Spec` is a _specification_ that describes the consensus engine. It was not designed to be used as a parent instance. Refactoring it in this way probably has unforseen consequences. `Engine` is meant to be the parent type to be passed around.